File: server-rails/app/models/scopes/task_scope.rb
# frozen_string_literal: true

module Scopes
  # TaskScope -- members only see tasks assigned to them.
  #
  # Auto-discovered by the HasAutoScope concern via naming convention.
  class TaskScope < AgentCode::ResourceScope
    def apply(relation)
      # Members only see tasks assigned to them
      if role == "member"
        relation.where(assignee_id: user.id)
      else
        relation
      end
    end
  end
end

File: server-nestjs/src/scopes/TaskScope.ts
import type { AgentCodeScope } from '@agentcode-dev/agentcode-nestjs';

/**
 * Restrict task listings so that "member" role users only see tasks assigned
 * to them. PRD §4 / AC-5.
 */
export class TaskScope implements AgentCodeScope {
  apply(where: Record<string, any>, ctx: { user?: any; userRole?: string | null }) {
    if (ctx.userRole === 'member' && ctx.user?.id) {
      return { ...where, assignedTo: ctx.user.id };
    }
    return where;
  }
}
